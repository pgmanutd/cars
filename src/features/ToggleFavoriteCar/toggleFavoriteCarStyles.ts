import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      border: `1px solid ${theme.palette.divider}`,
      padding: theme.spacing(3),
    },
    buttonContainer: {
      display: 'flex',
      justifyContent: 'flex-end',
      marginTop: theme.spacing(2),
    },
  }),
);
