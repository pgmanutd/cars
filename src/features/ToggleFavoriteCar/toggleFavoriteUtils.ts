import _omit from 'lodash/fp/omit';
import _curry from 'lodash/fp/curry';

import { Car } from 'features/CarInfo/types';

import { FavoriteCars } from './types';

export const addCarToFavorites = _curry(
  (car: Car, favoriteCars: FavoriteCars) => ({
    ...favoriteCars,
    [car.stockNumber]: car,
  }),
);

export const removeCarFromFavorites = _curry(
  (car: Car, favoriteCars: FavoriteCars) =>
    _omit([car.stockNumber], favoriteCars),
);

export const isCarInFavoriteCars = (car: Car, favoriteCars: FavoriteCars) =>
  favoriteCars[car.stockNumber];
