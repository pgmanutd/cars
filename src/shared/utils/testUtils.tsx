import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { render } from '@testing-library/react';

export const renderWithRouter = (
  ui: React.ReactElement,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  }: { route?: string; history?: MemoryHistory } = {},
) => ({
  ...render(<Router history={history}>{ui}</Router>),
  history,
});
