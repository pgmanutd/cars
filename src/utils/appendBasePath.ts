const appendBasePath = (basePath: string, path: string) =>
  `${basePath}?${path}`;

export default appendBasePath;
