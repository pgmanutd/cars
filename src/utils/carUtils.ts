import { Car } from 'features/CarInfo/types';

export const getCarName = ({ manufacturerName, modelName }: Car) =>
  `${manufacturerName} ${modelName}`;

export const getCarFeatures = (
  { stockNumber, mileage, fuelType, color }: Car,
  { stockNumberPrefix = '' }: { stockNumberPrefix?: string } = {
    stockNumberPrefix: '',
  },
) =>
  `${stockNumberPrefix}${stockNumber} - ${mileage?.number} ${mileage?.unit} - ${fuelType} - ${color}`;
