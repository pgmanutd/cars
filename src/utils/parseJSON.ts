const parseJSON = <T>(stringifiedJSON: string | null, fallbackValue: T) => {
  try {
    if (!stringifiedJSON) {
      return fallbackValue;
    }

    return JSON.parse(stringifiedJSON);
  } catch (e) {
    return fallbackValue;
  }
};

export default parseJSON;
