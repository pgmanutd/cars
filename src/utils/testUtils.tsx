import React from 'react';
import { Router, Route } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { HelmetProvider } from 'react-helmet-async';
import { render } from '@testing-library/react';

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    routerConfig: {
      path = '/',
      route = '/',
      history = createMemoryHistory({ initialEntries: [route] }),
    },
  }: {
    routerConfig: { path?: string; route?: string; history?: MemoryHistory };
  } = { routerConfig: {} },
) => ({
  ...render(
    <HelmetProvider>
      <Router history={history}>
        <Route
          path={path}
          render={routeProps => React.cloneElement(ui, { ...routeProps })}
        />
      </Router>
    </HelmetProvider>,
  ),
  history,
});
