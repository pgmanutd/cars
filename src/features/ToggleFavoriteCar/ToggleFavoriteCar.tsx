import React, { memo, useCallback } from 'react';
import Typography from '@material-ui/core/Typography';
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import _compose from 'lodash/fp/compose';

import { LOCAL_STORAGE_KEYS } from 'constants/appConstants';

import useLocalStorage from 'hooks/useLocalStorage';

import { useTranslate } from 'features/Translate';
import { Car } from 'features/CarInfo/types';

import { FavoriteCars } from './types';
import { useStyles } from './toggleFavoriteCarStyles';

import {
  isCarInFavoriteCars,
  addCarToFavorites,
  removeCarFromFavorites,
} from './toggleFavoriteUtils';

export interface ToggleFavoriteCarProps {
  car: Car;
}

const ToggleFavoriteCar: React.FC<ToggleFavoriteCarProps> = ({
  car,
  ...restProps
}) => {
  const classes = useStyles();
  const { translate } = useTranslate();
  const [favoriteCars, setFavoriteCars] = useLocalStorage<FavoriteCars>(
    LOCAL_STORAGE_KEYS.favoriteCars,
    {},
  );

  const isCarSetAsFavorite = isCarInFavoriteCars(car, favoriteCars);
  const toggleFavoriteCar = isCarSetAsFavorite
    ? removeCarFromFavorites(car)
    : addCarToFavorites(car);

  const handleFavoriteCarsClick = useCallback(() => {
    _compose(setFavoriteCars, toggleFavoriteCar)(favoriteCars);
  }, [toggleFavoriteCar, favoriteCars, setFavoriteCars]);

  return (
    <Box
      data-testid="ToggleFavoriteCar"
      {...restProps}
      className={classes.root}
    >
      <Typography variant="body2">
        {isCarSetAsFavorite
          ? translate('features.ToggleFavoriteCar.unFavoriteCarsDescription')
          : translate('features.ToggleFavoriteCar.favoriteCarsDescription')}
      </Typography>
      <Box className={classes.buttonContainer}>
        <Button
          variant="contained"
          color={isCarSetAsFavorite ? 'secondary' : 'primary'}
          onClick={handleFavoriteCarsClick}
        >
          {isCarSetAsFavorite
            ? translate('features.ToggleFavoriteCar.removeFavoriteButton')
            : translate('features.ToggleFavoriteCar.saveFavoriteButton')}
        </Button>
      </Box>
    </Box>
  );
};

export default memo(ToggleFavoriteCar);
