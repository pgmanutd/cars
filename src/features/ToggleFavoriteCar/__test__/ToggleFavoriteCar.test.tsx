import React from 'react';
import userEvent from '@testing-library/user-event';

import { renderWithTranslate } from 'utils/testUtils';

import ToggleFavoriteCar from '../ToggleFavoriteCar';

describe('<ToggleFavoriteCar />', () => {
  beforeEach(() => {
    localStorage.clear();
  });

  const setup = () => {
    const car = {
      stockNumber: 78924,
      manufacturerName: 'Fiat',
      modelName: 'Sedici',
      color: 'white',
      mileage: { number: 141474, unit: 'km' },
      fuelType: 'Diesel',
      pictureUrl: 'https://dummyimage.com/300/09f/fff.png',
    };
    const { renderResult } = renderWithTranslate(
      <ToggleFavoriteCar car={car} />,
    );

    return {
      renderResult,
    };
  };

  it('should render the component and matches it against stored snapshot', () => {
    const { renderResult } = setup();

    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it('should add car to local storage and show "Remove" button', () => {
    const { renderResult } = setup();

    userEvent.click(renderResult.getByText('Save'));

    expect(renderResult.getByText('Remove')).toBeInTheDocument();
  });

  it('should remove car from local storage and show "Save" button', () => {
    const { renderResult } = setup();

    userEvent.click(renderResult.getByText('Save'));
    userEvent.click(renderResult.getByText('Remove'));

    expect(renderResult.getByText('Save')).toBeInTheDocument();
  });
});
