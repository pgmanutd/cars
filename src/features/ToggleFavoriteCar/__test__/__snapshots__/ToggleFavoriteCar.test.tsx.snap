// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleFavoriteCar /> should render the component and matches it against stored snapshot 1`] = `
<DocumentFragment>
  <section
    class="MuiBox-root MuiBox-root-4 makeStyles-root-1"
    data-testid="ToggleFavoriteCar"
  >
    <p
      class="MuiTypography-root MuiTypography-body2"
    >
      If you like this car, click the button and save it in your collection of favorite items.
    </p>
    <div
      class="MuiBox-root MuiBox-root-35 makeStyles-buttonContainer-2"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Save
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </section>
</DocumentFragment>
`;
